File: src/cmd/simulate.go
package cmd

import (
	"fmt"
	"ordercontroller/controller"
	"time"

	"github.com/spf13/cobra"
)

var simulateCmd = &cobra.Command{
	Use:   "simulate",
	Short: "运行模拟场景",
	Long:  "运行固定的模拟场景，输出订单处理流程到 stdout",
	RunE: func(cmd *cobra.Command, args []string) error {
		runSimulate()
		return nil
	},
}

func init() {
	rootCmd.AddCommand(simulateCmd)
}

func runSimulate() {
	clk := controller.RealClock{}

	ctrl := controller.NewController(func(e controller.Event) {
		fmt.Println(controller.FormatEvent(e))
	}, clk)

	fmt.Println("McDonald's Order Management System - Simulation Results")
	fmt.Println()

	// 模拟场景
	ctrl.NewOrder(controller.OrderNormal) // #1000
	time.Sleep(1 * time.Second)

	ctrl.NewOrder(controller.OrderVIP) // #1001
	time.Sleep(0)

	ctrl.NewOrder(controller.OrderNormal) // #1002
	time.Sleep(1 * time.Second)

	ctrl.AddBot() // Bot #1 → 取走 VIP #1001
	time.Sleep(1 * time.Second)

	ctrl.AddBot() // Bot #2 → 取走 Normal #1000
	time.Sleep(1 * time.Second)

	// 等待 Bot #1 完成 VIP #1001 (10秒)
	time.Sleep(9 * time.Second)

	// 等待 Bot #2 完成 Normal #1000
	time.Sleep(1 * time.Second)

	// 创建新 VIP 订单
	ctrl.NewOrder(controller.OrderVIP) // #1003
	time.Sleep(1 * time.Second)

	// 等待剩余处理完成
	time.Sleep(10 * time.Second)

	// 移除 Bot #2
	ctrl.RemoveBot()
	time.Sleep(500 * time.Millisecond)

	// 等待最后的处理完成
	time.Sleep(10 * time.Second)

	// 打印最终状态
	status := ctrl.GetStatus()
	fmt.Println()
	fmt.Println("Final Status:")
	fmt.Printf("- Total Orders Processed: %d\n", status.TotalOrders)
	fmt.Printf("- Orders Completed: %d\n", status.Complete)
	fmt.Printf("- Active Bots: %d\n", status.ActiveBots+status.IdleBots)
	fmt.Printf("- Pending Orders: %d\n", status.Pending)
}

File: src/simulate.ts
import { OrderController } from "./domain/OrderController.js";
import { createTimestampedLogger } from "./domain/logger.js";

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function runSimulation(): Promise<void> {
  console.log("McDonald's Order Management System - Simulation Results\n");
  const logger = createTimestampedLogger();
  const controller = new OrderController({ logger });

  // Requirement 1 & 2: create Normal then VIP then Normal orders.
  // VIP must place ahead of all Normal orders but behind existing VIP.
  controller.createOrder("NORMAL");
  controller.createOrder("VIP");
  controller.createOrder("NORMAL");

  // Requirement 4: add a bot; it immediately picks the highest-priority pending order.
  controller.addBot();
  controller.addBot();

  // Wait for the first batch to finish (real 10s processing time).
  await sleep(10_000);

  // Requirement 5: after completing, bots pick up remaining pending orders
  // or become IDLE. Now create a VIP order; an idle bot should pick it up.
  controller.createOrder("VIP");

  // Requirement 6: while both bots are busy, create a Normal order so it stays
  // pending, then add a bot that immediately picks it up. Destroy that newest
  // bot mid-process; the order returns to pending and is picked up once a bot
  // becomes free. Do this before the sleep so no bot is idle yet.
  controller.createOrder("NORMAL");
  controller.addBot();
  controller.removeBot();

  // Wait for the in-progress orders to finish and the returned order to complete.
  await sleep(10_000);

  // Wait for the re-picked-up order (#1005) to finish processing.
  await sleep(10_000);

  // Requirement: manager decreases bots; destroy an idle bot.
  controller.removeBot();

  console.log("");
  console.log("Final Status:");
  const total = controller.completedCount + controller.pendingCount;
  const vip = controller.vipCompletedCount();
  const normal = controller.normalCompletedCount();
  console.log(`- Total Orders Processed: ${total} (${vip} VIP, ${normal} Normal)`);
  console.log(`- Orders Completed: ${controller.completedCount}`);
  console.log(`- Active Bots: ${controller.botCount}`);
  console.log(`- Pending Orders: ${controller.pendingCount}`);
}
